### ISSUE TITLE
export various unexposed types

### ISSUE THREAD

**Issue by cdaringe** [octokit/plugin-rest-endpoint-methods.js#47]
**What’s missing?**

many types, such as `SearchCommitsResponse` are _not exposed_.  alternatively, perhapes we consider enhanced type inference on `.paginate(...)` to infer expected output types from inputs

**Why?**

consider the following:

`const res = await octokit.paginate(searchCommitOptions) // PaginationResults<unknown>`

it would be great have better typing here.  it is constrained because it expects a generic--`<T>(options: OctokitTypes.EndpointOptions): Promise<PaginationResults<T>>;`--and i cannot pass `SearchCommitsResponse` in, because it's not exported.

**Alternatives you tried**

n/a

**Comment by gr2m** [comment]
I will export response types in `@octokit/types`. There was a similar issue reported here:
https://github.com/octokit/rest.js/issues/1624#issuecomment-590388141, there are workarounds for the time being mentioned in the thread for the time being. Does that workaround work for you?

**Comment by cdaringe** [comment]
thanks @gr2m. ya, those type unpackers where what i was teasing.  no rush--thx for the attention

**Comment by hipstersmoothie** [comment]
I fully second the better types around pagination. 

Currently kinda blocked on upgrading here (https://github.com/intuit/auto/pull/1003). We heavily use the types available in 16 and doing those typescript tricks is a band-aid I don't really want to commit.

**Comment by gr2m** [comment]
I'm not sure if exporting types is the baind-aid.

In an ideal world, you shouldn't need to set any types at all, something like this should just work

```js
const results = await octokit.paginate(octokit.search.commits, { q })
```

Or this

```js
const results = await octokit.paginate("GET /search/commits", { q })
```

Looking up the type after you looked up the endpoint method is double the work. I still think we should export the response types from `@octokit/types`, but it's not the right solution for the problem with `octokit.paginate`, I think

**Comment by gr2m** [comment]
I've an open pull request for `@octokit/types` that will add two helpers

1. GetResponseType
2. GetResponseDataType

https://github.com/octokit/types.ts/pull/32

These allow to extract the response type / response data type from the endpoint methods.

Example:

```ts
import { GetResponseType, GetResponseDataType } from "@octokit/types";
import { Octokit } from "@octokit/rest";

const octokit = new Octokit();
type CreateLabelResponseType = GetResponseType<
  typeof octokit.issues.createLabel
>;
type CreateLabelResponseDataType = GetResponseDataType<
  typeof octokit.issues.createLabel
>;
```

I will create a separate pull request which will export a type for each response type in a separate PR. I'm not yet sure if I'll export the response type in `@octokit/types` or `@octokit/plugin-rest-endpoint-methods`, or both

I'll also look into implementing a proper solution for `octokit.paginate`. The response data types don't always match what `octokit.paginate` gives you, because some responses are different than others and are being normalized within `octokit.paginate`, see https://github.com/octokit/plugin-paginate-rest.js#how-it-works

Feedback welcome 😁

**Comment by cdaringe** [comment]
sweet!

**Comment by hipstersmoothie** [comment]
Once you get a release out I can help with testing

**Comment by gr2m** [comment]
@hipstersmoothie I've published a beta preview, you can install it with 

```
npm install @octokit/types@beta
# should install 2.8.0-beta.1
```

**Comment by hipstersmoothie** [comment]
Types are looking better! Running into an error originating in @octokit/types right now though

![Screen Shot 2020-04-13 at 9 53 06 AM](https://user-images.githubusercontent.com/1192452/79140289-994b2480-7d6c-11ea-8bfb-851e397dd809.png)

**Comment by hipstersmoothie** [comment]
The only thing left to update `auto` to v17 of octokit is for me to figure out how to paginate some calls now.

This is how I was doing it before:

```ts
const labels: Octokit.IssuesListLabelsForRepoResponse = await this.github.paginate(
  this.github.issues.listLabelsForRepo.endpoint(args)
);
```

**Comment by gr2m** [comment]
can you try `this.github.issues.listLabelsForRepo.endpoint(args) as EndpointOptions` and `import { EndpointOptions } from "@octokit/types"`?

I'm working on a better solution to the problem right now: https://github.com/octokit/plugin-paginate-rest.js/pull/29

**Comment by gr2m** [comment]
@cdaringe regarding your initial issue, I'm working on a pull request right now at https://github.com/octokit/plugin-rest-endpoint-methods.js/pull/65 which will start exporting `RestEndpointMethodTypes`, so you will be able to access parameter and response types

```ts
RestEndpointMethodTypes["issues"]["updateLabel"]["parameters"]
RestEndpointMethodTypes["issues"]["updateLabel"]["response"]
```

**Comment by hipstersmoothie** [comment]
Didn't work

![Screen Shot 2020-04-13 at 2 00 01 PM](https://user-images.githubusercontent.com/1192452/79160703-12f40a00-7d8f-11ea-99bf-a21d2c12f207.png)

**Comment by hipstersmoothie** [comment]
Didn't work

![Screen Shot 2020-04-13 at 2 01 52 PM](https://user-images.githubusercontent.com/1192452/79160847-551d4b80-7d8f-11ea-9b30-469fc0937c0a.png)

**Comment by gr2m** [comment]
Could you share a minimal test case to reproduce the problem?

**Comment by hipstersmoothie** [comment]
https://github.com/hipstersmoothie/octokit-repro
